library/alloc/src/vec.rs

```diff
@@ -2391,11 +2384,62 @@ where
 }
 
 impl<T> SpecExtend<T, IntoIter<T>> for Vec<T> {
-    fn spec_extend(&mut self, mut iterator: IntoIter<T>) {
-        unsafe {
-            self.append_elements(iterator.as_slice() as _);
+    fn spec_extend(&mut self, iterator: IntoIter<T>) {
+        // Avoid reallocation if we can use iterator's storage instead. This requires 1 memcpy and 0-1 memmove
+        // while reallocation would require 1 alloc, 1-2 memcpy, 1-2 free.
+        //
+        // # illustration of some extend scenarios (not exhaustive)
+        //
+        //   == step ==              == memory ==              == self ==         == iter / v ==
+        //                   0123456789abcdef0123456789abcdef
+        //                   0---------------1---------------
+        //
+        // ## non-empty self, partially consumed iterator
+        //
+        //    [initial]      AAAA_-----__BBB___--------------  Vec(0x00, 4, 5)    IntoIter(0x0a, 0x0c, 0x0f, 8)
+        // ³  into_vec       AAAA_-----____BBB_--------------  Vec(0x00, 4, 5)    Vec(0x0a, 7, 8)
+        // ²  prepend        _____-----AAAABBB_--------------  Vec(0x00, 0, 5)    Vec(0x0a, 7, 8)
+        // ⁴  *self = v      ----------AAAABBB_--------------  Vec(0x0a, 7, 8)
+        //
+        // ## empty self, partially consumed iterator
+        //
+        //    [initial]      ____------__BBBB__--------------  Vec(0x00, 0, 4)    IntoIter(0x0a, 0x0c, 0x10, 8)
+        // ³  into_vec       ____------BBBB____--------------  Vec(0x00, 0, 4)    Vec(0x0a, 4, 8)
+        // ⁴  *self = v      ----------BBBB____--------------  Vec(0x0a, 4, 8)
+        //
+        // ## empty self, pristine iterator
```

Comment:
Do we even check for this case below?

By the way, nice diagram.

Reply:
`into_vec_with_uninit_prefix` is pretty much just a struct conversion when there's nothing to move.